### PR TITLE
Draft: add tests for transforms

### PR DESCRIPTION
**Pull request by spersson**
Addressing issue #635.
I was thinking the easiest is to add round trip tests. But I'm not sure if it's good enough? It's a start at least. Should I also hardcode some cases and expected results?
I was also not sure about generating different random numbers on every test run. So far it's working, but perhaps there's a policy.
I commented out the types and sizes that are not supported now. If there are no plans to add support I can just remove those lines.
Tolerance of +-1 is accepted now, tolerance of 0 didn't work for bigger block sizes.
This is my first time writing tests for software, despite having tested software for a living over ten years now... :)

### CODE REVIEW

**Comment by lu-zero** [comment]
Another option is to compare implementations.

The rust implementation can be compared with the libaom one and the optimized implementations compared with the rust one once validated.

Any new code relying on the libaom copy should be optional since we plan to move away from aom.

**Comment by ycho** [comment]
I also observed that there are errors of +-1 for 32x32 tx, and identity tx shows errors at even smaller size tx. For 32x32, we view that it is expected since less # of fractional bits are used, i.e. 2 bits compared to 3 bits for 4x4, 8x8, and 16x16.

**Comment by tdaede** [comment]
Roundtrip tests are good. While the inverse transforms are required to be bitexact, the forward ones aren't, so we have a bit of leeway in implementation detail, and the +1/-1 thresholds are good for that.

If you'd like to add more involved dct tests, there's good examples of them in daala's dct.c file: http://git.xiph.org/?p=daala.git;a=blob;f=src/dct.c;h=adb0faf549bc9d164ae218c7cb085c6d56b90dbd;hb=HEAD

File: src/transform/mod.rs [hunk 1]
```diff
@@ -276,3 +276,88 @@ pub fn inverse_transform_add(
...
+    for (s, d) in src.iter().zip(dst) {
+      assert!(i16::abs((*s as i16) - (*d as i16)) <= 1);
+    }
+  }
+
+  #[test]
+  fn roundtrips() {
+    use partition::TxSize::*;
+    use partition::TxType::*;
+    let combinations = vec![
```
Comment (tdaede) [inline src/transform/mod.rs]: This can just be a static array.

Comment (spersson) [reply src/transform/mod.rs]: I had tried that but had problem with IntoIter trait not implemented for arrays larger than 32. I looked some more now and it seems that the iter method works for all sizes. I will push this change soon.

**Comment by spersson** [comment]
Thanks for the feedback, all of you! I just made a change to use smallest possible tolerances that still passes the test. A little bit worrying is that by running the tests many times I found that DCT_DCT for blocksize 32 sometimes gave error -+2.

**Comment by tdaede** [comment]
OK. I'll land it with the +/-2 limits.

**Pull request merged**